src/main/java/com/puppycrawl/tools/checkstyle/filters/XpathFilter.java

```diff
@@ -0,0 +1,184 @@
+import com.puppycrawl.tools.checkstyle.xpath.AbstractNode;
+import com.puppycrawl.tools.checkstyle.xpath.RootNode;
+import net.sf.saxon.om.Item;
+import net.sf.saxon.sxpath.XPathDynamicContext;
+import net.sf.saxon.sxpath.XPathEvaluator;
+import net.sf.saxon.sxpath.XPathExpression;
+import net.sf.saxon.trans.XPathException;
+
+/**
+ * This filter processes {@link TreeWalkerAuditEvent}
+ * objects based on the criteria of file, check, module id, xpathQuery.
+ *
+ * @author Timur Tibeyev.
+ */
+public class XpathFilter implements TreeWalkerFilter {
+    /** The regexp to match file names against. */
+    private final Pattern fileRegexp;
+
+    /** The pattern for file names. */
+    private final String filePattern;
+
+    /** The regexp to match check names against. */
+    private final Pattern checkRegexp;
+
+    /** The pattern for check class names. */
+    private final String checkPattern;
+
+    /** Module id filter. */
+    private final String moduleId;
+
+    /** Xpath expression. */
+    private final XPathExpression xpathExpression;
+
+    /** Xpath query. */
+    private final String xpathQuery;
+
+    /**
+     * Creates a {@code XpathElement} instance.
+     * @param files regular expression for names of filtered files
+     * @param checks regular expression for filtered check classes
+     * @param moduleId the module id
+     * @param query the xpath query
+     */
+    public XpathFilter(String files, String checks,
+                       String moduleId, String query) {
+        filePattern = files;
+        fileRegexp = Pattern.compile(files);
+        checkPattern = checks;
+        if (checks == null) {
+            checkRegexp = null;
+        }
+        else {
+            checkRegexp = CommonUtils.createPattern(checks);
+        }
+        this.moduleId = moduleId;
+        xpathQuery = query;
+        if (xpathQuery == null) {
+            xpathExpression = null;
+        }
+        else {
+            final XPathEvaluator xpathEvaluator = new XPathEvaluator();
+            try {
+                xpathExpression = xpathEvaluator.createExpression(xpathQuery);
+            }
+            catch (XPathException ex) {
+                throw new IllegalStateException("Unexpected xpath query", ex);
+            }
+        }
+    }
+
+    @Override
+    public boolean accept(TreeWalkerAuditEvent event) {
+        return !isFileNameAndModuleAndCheckNameMatching(event)
+                || !isXpathQueryMatching(event);
+    }
+
+    /**
+     * Is matching by file name, moduleId and Check name.
+     * @param event event
+     * @return true if it is matching
+     */
+    private boolean isFileNameAndModuleAndCheckNameMatching(TreeWalkerAuditEvent event) {
+        return event.getFileName() != null
+                && fileRegexp.matcher(event.getFileName()).find()
+                && event.getLocalizedMessage() != null
+                && (moduleId == null || moduleId.equals(event.getModuleId()))
```

Comment:
This should be a NOT of https://github.com/checkstyle/checkstyle/blob/f244f7f1e2fe49d2fcda9cd5c901d751769f73e8/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressElement.java#L120-L125 .
It looks correct to me.

> If either moduleId or checkRegexp match the event's module id or check, than the audit event should be denied.

This doesn't sound correct to me. Each element of a suppression is an `AND` not an `OR`.
If you supply module id and check regexp, the event shouldn't be suppressed if only one matches. It should only be suppressed if both match.

> I think, that is should be:
> `|| checkRegexp != null && checkRegexp.matcher(event.getSourceName()).find())`

This doesn't seem right to me.
As I mentioned before, it should be a complete NOT of https://github.com/checkstyle/checkstyle/blob/f244f7f1e2fe49d2fcda9cd5c901d751769f73e8/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressElement.java#L120-L125 .
What I have seen so far, only `accept` method isn't coded correctly.